File: integrations/go/conformance.go
package main

import (
	"bytes"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

type vectors struct {
	Valid   []vector `json:"valid"`
	Invalid []vector `json:"invalid"`
}

type vector struct {
	Name             string         `json:"name"`
	Wire             map[string]any `json:"wire"`
	CanonicalJSON    string         `json:"canonical_json"`
	CanonicalMsgpack string         `json:"canonical_msgpack_hex"`
	CanonicalSHA     string         `json:"canonical_sha256"`
}

func prefix(code byte, value uint64, n int) []byte {
	out := make([]byte, 1+n)
	out[0] = code
	switch n {
	case 1:
		out[1] = byte(value)
	case 2:
		binary.BigEndian.PutUint16(out[1:], uint16(value))
	case 4:
		binary.BigEndian.PutUint32(out[1:], uint32(value))
	case 8:
		binary.BigEndian.PutUint64(out[1:], value)
	}
	return out
}

func encInt(v int64) []byte {
	if v >= 0 {
		u := uint64(v)
		if u <= 0x7f {
			return []byte{byte(u)}
		}
		if u <= 0xff {
			return prefix(0xcc, u, 1)
		}
		if u <= 0xffff {
			return prefix(0xcd, u, 2)
		}
		if u <= 0xffffffff {
			return prefix(0xce, u, 4)
		}
		return prefix(0xcf, u, 8)
	}
	if v >= -32 {
		return []byte{byte(256 + v)}
	}
	if v >= -128 {
		return []byte{0xd0, byte(int8(v))}
	}
	if v >= -32768 {
		out := []byte{0xd1, 0, 0}
		binary.BigEndian.PutUint16(out[1:], uint16(int16(v)))
		return out
	}
	if v >= -2147483648 {
		out := make([]byte, 5)
		out[0] = 0xd2
		binary.BigEndian.PutUint32(out[1:], uint32(int32(v)))
		return out
	}
	return prefix(0xd3, uint64(v), 8)
}

func encString(s string) []byte {
	b := []byte(s)
	var h []byte
	n := len(b)
	if n <= 31 {
		h = []byte{byte(0xa0 | n)}
	} else if n <= 0xff {
		h = prefix(0xd9, uint64(n), 1)
	} else if n <= 0xffff {
		h = prefix(0xda, uint64(n), 2)
	} else {
		h = prefix(0xdb, uint64(n), 4)
	}
	return append(h, b...)
}

func encode(v any) ([]byte, error) {
	switch x := v.(type) {
	case nil:
		return []byte{0xc0}, nil
	case bool:
		if x {
			return []byte{0xc3}, nil
		}
		return []byte{0xc2}, nil
	case string:
		return encString(x), nil
	case json.Number:
		if !strings.ContainsAny(string(x), ".eE") {
			if i, e := x.Int64(); e == nil {
				return encInt(i), nil
			}
		}
		f, e := x.Float64()
		if e != nil || math.IsNaN(f) || math.IsInf(f, 0) {
			return nil, errors.New("invalid number")
		}
		out := make([]byte, 9)
		out[0] = 0xcb
		binary.BigEndian.PutUint64(out[1:], math.Float64bits(f))
		return out, nil
	case []any:
		n := len(x)
		var h []byte
		if n <= 15 {
			h = []byte{byte(0x90 | n)}
		} else if n <= 0xffff {
			h = prefix(0xdc, uint64(n), 2)
		} else {
			h = prefix(0xdd, uint64(n), 4)
		}
		out := bytes.NewBuffer(h)
		for _, item := range x {
			b, e := encode(item)
			if e != nil {
				return nil, e
			}
			out.Write(b)
		}
		return out.Bytes(), nil
	case map[string]any:
		keys := make([]string, 0, len(x))
		for k := range x {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		n := len(keys)
		var h []byte
		if n <= 15 {
			h = []byte{byte(0x80 | n)}
		} else if n <= 0xffff {
			h = prefix(0xde, uint64(n), 2)
		} else {
			h = prefix(0xdf, uint64(n), 4)
		}
		out := bytes.NewBuffer(h)
		for _, k := range keys {
			out.Write(encString(k))
			b, e := encode(x[k])
			if e != nil {
				return nil, e
			}
			out.Write(b)
		}
		return out.Bytes(), nil
	default:
		return nil, fmt.Errorf("unsupported type %T", v)
	}
}

func canonicalJSON(v any) ([]byte, error) { return json.Marshal(v) }

var traceRe = regexp.MustCompile(`^[0-9a-f]{2}-[0-9a-f]{32}-[0-9a-f]{16}-[0-9a-f]{2}$`)

func validate(w map[string]any) error {
	allowed := map[string]bool{"v": true, "c": true, "a": true, "i": true, "s": true, "r": true, "x": true, "R": true, "b": true, "d": true, "p": true, "m": true, "g": true, "z": true}
	for k := range w {
		if !allowed[k] {
			return fmt.Errorf("unknown top-level key %s", k)
		}
	}
	vn, ok := w["v"].(json.Number)
	if !ok || string(vn) != "2" {
		return errors.New("wire version")
	}
	if _, ok := w["c"].(string); !ok {
		return errors.New("codebook")
	}
	if p, ok := w["p"].(map[string]any); ok {
		if q, exists := p["q"]; exists {
			n, ok := q.(json.Number)
			if !ok {
				return errors.New("provenance confidence")
			}
			f, _ := n.Float64()
			if f < 0 || f > 1 {
				return errors.New("provenance confidence")
			}
		}
	}
	if raw, exists := w["x"]; exists {
		arr, ok := raw.([]any)
		if !ok {
			return errors.New("atoms")
		}
		atomAllowed := map[string]bool{"c": true, "v": true, "l": true, "h": true, "p": true, "q": true, "e": true}
		for _, a := range arr {
			m, ok := a.(map[string]any)
			if !ok {
				return errors.New("atom")
			}
			for k := range m {
				if !atomAllowed[k] {
					return errors.New("unknown atom key")
				}
			}
			if q, exists := m["q"]; exists {
				n, ok := q.(json.Number)
				if !ok {
					return errors.New("atom confidence")
				}
				f, _ := n.Float64()
				if f < 0 || f > 1 {
					return errors.New("atom confidence")
				}
			}
			if e, exists := m["e"]; exists {
				if _, ok := e.(string); !ok {
					return errors.New("epistemic type")
				}
			}
		}
	}
	if g, ok := w["g"].(map[string]any); ok {
		if g["alg"] != "Ed25519" {
			return errors.New("signature algorithm")
		}
	}
	if z, ok := w["z"].(map[string]any); ok {
		p, ok := z["p"].(string)
		if !ok || !traceRe.MatchString(p) {
			return errors.New("traceparent")
		}
		if s, exists := z["s"]; exists {
			if _, ok := s.(string); !ok {
				return errors.New("tracestate")
			}
		}
	}
	return nil
}

func main() {
	path := "tck/vectors/core.json"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}
	data, err := os.ReadFile(path)
	if err != nil {
		panic(err)
	}
	dec := json.NewDecoder(bytes.NewReader(data))
	dec.UseNumber()
	var suite vectors
	if err := dec.Decode(&suite); err != nil {
		panic(err)
	}
	failures := []string{}
	total := 0
	for _, t := range suite.Valid {
		total++
		if err := validate(t.Wire); err != nil {
			failures = append(failures, t.Name+": "+err.Error())
			continue
		}
		cj, _ := canonicalJSON(t.Wire)
		packed, e := encode(t.Wire)
		if e != nil {
			failures = append(failures, t.Name+": "+e.Error())
			continue
		}
		sum := sha256.Sum256(packed)
		if t.CanonicalJSON != "" && string(cj) != t.CanonicalJSON {
			failures = append(failures, t.Name+": canonical JSON mismatch")
		}
		if hex.EncodeToString(packed) != t.CanonicalMsgpack {
			failures = append(failures, t.Name+": canonical MessagePack mismatch")
		}
		if "sha256:"+hex.EncodeToString(sum[:]) != t.CanonicalSHA {
			failures = append(failures, t.Name+": digest mismatch")
		}
	}
	for _, t := range suite.Invalid {
		total++
		if validate(t.Wire) == nil {
			failures = append(failures, t.Name+": invalid vector accepted")
		}
	}
	result := map[string]any{"ok": len(failures) == 0, "total": total, "passed": total - len(failures), "failures": failures}
	out, _ := json.Marshal(result)
	fmt.Println(string(out))
	if len(failures) > 0 {
		os.Exit(1)
	}
	_ = strconv.IntSize
}

File: integrations/openclaw/src/conformance.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// SAGE is dual-licensed under AGPL-3.0 and a commercial license.
// Contact sage@digitalacre.org for commercial licensing.
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";

/**
 * SAGE JavaScript TCK conformance runner.
 *
 * Implements the SAGE wire-format canonicalization, MessagePack-style
 * encoding, and validation rules, then verifies every vector in the TCK
 * corpus. Used by `npm run tck` and the three-runtime conformance matrix.
 */

type WireValue = null | boolean | number | string | WireValue[] | { [key: string]: WireValue };

type Vector = {
  name: string;
  wire: Record<string, WireValue>;
  canonical_json?: string;
  canonical_msgpack_hex?: string;
  canonical_sha256?: string;
};

type Corpus = {
  protocol: string;
  wire_version: number;
  valid: Vector[];
  invalid: Vector[];
};

function normalize(value: WireValue): WireValue {
  if (Array.isArray(value)) return value.map(normalize);
  if (value && typeof value === "object" && !Array.isArray(value)) {
    return Object.fromEntries(
      Object.keys(value as Record<string, WireValue>)
        .sort()
        .map((key) => [key, normalize((value as Record<string, WireValue>)[key])]),
    );
  }
  if (typeof value === "number" && !Number.isFinite(value)) throw new Error("non-finite number");
  return value;
}

function prefix(code: number, value: number, bytes: 1 | 2 | 4): Buffer {
  const out = Buffer.alloc(1 + bytes);
  out[0] = code;
  if (bytes === 1) out.writeUInt8(value, 1);
  else if (bytes === 2) out.writeUInt16BE(value, 1);
  else out.writeUInt32BE(value, 1);
  return out;
}

function encodeInteger(value: number): Buffer {
  if (value >= 0) {
    if (value <= 0x7f) return Buffer.from([value]);
    if (value <= 0xff) return prefix(0xcc, value, 1);
    if (value <= 0xffff) return prefix(0xcd, value, 2);
    if (value <= 0xffffffff) return prefix(0xce, value, 4);
    const out = Buffer.alloc(9);
    out[0] = 0xcf;
    out.writeBigUInt64BE(BigInt(value), 1);
    return out;
  }
  if (value >= -32) return Buffer.from([0x100 + value]);
  if (value >= -128) {
    const out = Buffer.alloc(2);
    out[0] = 0xd0;
    out.writeInt8(value, 1);
    return out;
  }
  if (value >= -32768) {
    const out = Buffer.alloc(3);
    out[0] = 0xd1;
    out.writeInt16BE(value, 1);
    return out;
  }
  if (value >= -2147483648) {
    const out = Buffer.alloc(5);
    out[0] = 0xd2;
    out.writeInt32BE(value, 1);
    return out;
  }
  const out = Buffer.alloc(9);
  out[0] = 0xd3;
  out.writeBigInt64BE(BigInt(value), 1);
  return out;
}

function encodeString(value: string): Buffer {
  const body = Buffer.from(value, "utf8");
  let head: Buffer;
  if (body.length <= 31) head = Buffer.from([0xa0 | body.length]);
  else if (body.length <= 0xff) head = prefix(0xd9, body.length, 1);
  else if (body.length <= 0xffff) head = prefix(0xda, body.length, 2);
  else head = prefix(0xdb, body.length, 4);
  return Buffer.concat([head, body]);
}

function encode(value: WireValue): Buffer {
  if (value === null) return Buffer.from([0xc0]);
  if (value === false) return Buffer.from([0xc2]);
  if (value === true) return Buffer.from([0xc3]);
  if (typeof value === "number") {
    if (Number.isSafeInteger(value)) return encodeInteger(value);
    const out = Buffer.alloc(9);
    out[0] = 0xcb;
    out.writeDoubleBE(value, 1);
    return out;
  }
  if (typeof value === "string") return encodeString(value);
  if (Array.isArray(value)) {
    const head =
      value.length <= 15
        ? Buffer.from([0x90 | value.length])
        : value.length <= 0xffff
          ? prefix(0xdc, value.length, 2)
          : prefix(0xdd, value.length, 4);
    return Buffer.concat([head, ...value.map(encode)]);
  }
  if (value && typeof value === "object") {
    const entries = Object.entries(normalize(value) as Record<string, WireValue>);
    const head =
      entries.length <= 15
        ? Buffer.from([0x80 | entries.length])
        : entries.length <= 0xffff
          ? prefix(0xde, entries.length, 2)
          : prefix(0xdf, entries.length, 4);
    return Buffer.concat([head, ...entries.flatMap(([key, item]) => [encodeString(key), encode(item)])]);
  }
  throw new Error(`unsupported type: ${typeof value}`);
}

function validate(wire: Record<string, WireValue>): void {
  const top = new Set(["v", "c", "a", "i", "s", "r", "x", "R", "b", "d", "p", "m", "g", "z"]);
  if (!wire || typeof wire !== "object" || Array.isArray(wire)) throw new Error("wire object required");
  if (wire.v !== 2) throw new Error("wire version");
  if (typeof wire.c !== "string") throw new Error("codebook");
  for (const key of Object.keys(wire)) if (!top.has(key)) throw new Error("unknown top-level key");
  const provenance = wire.p as Record<string, WireValue> | undefined;
  if (provenance?.q !== undefined && (typeof provenance.q !== "number" || provenance.q < 0 || provenance.q > 1)) {
    throw new Error("provenance confidence");
  }
  if (wire.x !== undefined) {
    if (!Array.isArray(wire.x)) throw new Error("atoms");
    const atomKeys = new Set(["c", "v", "l", "h", "p", "q", "e"]);
    for (const atom of wire.x as Record<string, WireValue>[]) {
      for (const key of Object.keys(atom)) if (!atomKeys.has(key)) throw new Error("unknown atom key");
      if (atom.q !== undefined && (typeof atom.q !== "number" || atom.q < 0 || atom.q > 1)) throw new Error("atom confidence");
      if (atom.e !== undefined && typeof atom.e !== "string") throw new Error("epistemic type");
    }
  }
  const sig = wire.g as Record<string, WireValue> | undefined;
  if (sig !== undefined && sig.alg !== "Ed25519") throw new Error("signature algorithm");
  const trace = wire.z as Record<string, WireValue> | undefined;
  if (trace !== undefined) {
    if (typeof trace.p !== "string" || !/^[0-9a-f]{2}-[0-9a-f]{32}-[0-9a-f]{16}-[0-9a-f]{2}$/.test(trace.p)) {
      throw new Error("traceparent");
    }
    if (trace.s !== undefined && typeof trace.s !== "string") throw new Error("tracestate");
  }
}

const file = process.argv[2] ?? new URL("../tck/core.json", import.meta.url);
const vectors = JSON.parse(readFileSync(file, "utf8")) as Corpus;
const failures: string[] = [];
let total = 0;
for (const test of vectors.valid) {
  total += 1;
  try {
    validate(test.wire);
    const canonical = normalize(test.wire);
    const json = JSON.stringify(canonical);
    const packed = encode(canonical);
    const digest = `sha256:${createHash("sha256").update(packed).digest("hex")}`;
    if (typeof test.canonical_json === "string" && json !== test.canonical_json) throw new Error("canonical JSON mismatch");
    if (packed.toString("hex") !== test.canonical_msgpack_hex) throw new Error("canonical MessagePack mismatch");
    if (digest !== test.canonical_sha256) throw new Error("digest mismatch");
  } catch (error) {
    failures.push(`${test.name}: ${error instanceof Error ? error.message : String(error)}`);
  }
}
for (const test of vectors.invalid) {
  total += 1;
  try {
    validate(test.wire);
    failures.push(`${test.name}: invalid vector accepted`);
  } catch {
    // expected rejection
  }
}
const result = { ok: failures.length === 0, total, passed: total - failures.length, failures };
console.log(JSON.stringify(result));
if (!result.ok) process.exitCode = 1;
